### PR TITLE
feat: convert TAO Audio PCI to IMS

### PR DESCRIPTION
**Pull request by bziondik**
Related to: 
https://oat-sa.atlassian.net/browse/TR-3906
https://oat-sa.atlassian.net/browse/TR-3909
https://oat-sa.atlassian.net/browse/TR-3910

## Technical Acceptance criteria

AC1. An integration branch of the extension [GitHub - oat-sa/extension-tao-itemqti-pci](https://github.com/oat-sa/extension-tao-itemqti-pci) contains the the files and folders for the IMS version of the Audio Recording PCI into [extension-tao-itemqti-pci/views/js/pciCreator/ims at master · oat-sa/extension-tao-itemqti-pci](https://github.com/oat-sa/extension-tao-itemqti-pci/tree/master/views/js/pciCreator/ims)

AC2. The source of the OAT Audio Recording PCI are migrated to the IMS version (the migration of the worker is excluded)

AC3. The manifest is created according to [taohub-articles/convert-tao-pci-to-ims-pci.md at master · oat-sa/taohub-articles](https://github.com/oat-sa/taohub-articles/blob/master/faq/convert-tao-pci-to-ims-pci.md)

AC4. The OAT version of the Audio Recording PCI is kept

AC1. The IMS version of the Audio Recording PCI does not contain the use media stimulus feature in the authoring

AC2. The IMS version of the Audio Recording PCI does not contain the use media stimulus feature in the runtime

AC3. All dependencies required from the  use media stimulus can be removed 

AC1. Any unused dependency of in the IMS Audio Recording PCI is removed

AC1a. If a dependency is used but can be safely replaced (lodash, jquery, etc.) it is replaced

AC1b. Direct dependencies to TAO Terre are remove (i18n) 

 AC2. The IMS Audio Recording PCI can be bundled according to the IMS standard (see [taohub-articles/convert-tao-pci-to-ims-pci.md at master · oat-sa/taohub-articles](https://github.com/oat-sa/taohub-articles/blob/master/faq/convert-tao-pci-to-ims-pci.md#bundle-pci) )

## What was done 
according https://github.com/oat-sa/taohub-articles/blob/master/faq/convert-tao-pci-to-ims-pci.md:
1. copied files from `qtiItemPci/views/js/pciCreator/dev/audioRecordingInteraction` to `qtiItemPci/views/js/pciCreator/ims/audioRecordingInteraction`
2. renamed `pciCreator.js` to `imsPciCreator.js`
3. renamed `pciCreator.json` to `imsPciCreator.json`
4. updated manifest  - removed `libraries` , `stylesheets`, `mediaFiles` from runtime.
5. split `"./runtime/img/controls.svg"` to 4 icons to require  them in main runtime, instead load by link
6. required files in runtime hook 
`qtiItemPci/views/js/pciCreator/ims/audioRecordingInteraction/runtime/audioRecordingInteraction.js`
```
'text!audioRecordingInteraction/runtime/img/play.svg',
    'text!audioRecordingInteraction/runtime/img/record.svg',
    'text!audioRecordingInteraction/runtime/img/reset.svg',
    'text!audioRecordingInteraction/runtime/img/stop.svg',
    'css!audioRecordingInteraction/runtime/css/audioRecordingInteraction',
    'audioRecordingInteraction/runtime/js/workers/WebAudioRecorderWav',
    'audioRecordingInteraction/runtime/js/workers/WavAudioEncoder'
```
7. remove handling of the prompt
8. replaced shared libraries to their equivalent in portableLib
9. [removed i18n](https://github.com/oat-sa/extension-tao-itemqti-pci/pull/294/commits/43405c7ce5ea81ab057f2d18dcfbd89a52c727f4) and support of translations
10. adapted PCI runtime API to IMS V1 specifications - added `typeIdentifier`, `getInstance`, `getState`, `oncompleted` methods
11. added scss file to sass task
12. [removed 'use media stimulus' feature](https://github.com/oat-sa/extension-tao-itemqti-pci/pull/294/commits/46568abfd5fbd6fb5944dd35bba31ca1394ba08d) because this feature related to code that has complex dependencies to TAO code and prevent to run PCI in Solar TR
13. [convert audioRecordingInteraction to factory](https://github.com/oat-sa/extension-tao-itemqti-pci/pull/294/commits/5dec93ee2153d0e16fc83aea005143021037952a) for each interaction should be created new Object because audioRecordingInteraction stores controls, handle events
14. refactored [modal jquery plugin](https://github.com/oat-sa/extension-tao-itemqti-pci/pull/294/commits/6fa5112471bed70e7a7f92bf8c7d3371987bd2b2)
15. [added unit test](https://github.com/oat-sa/extension-tao-itemqti-pci/pull/294/commits/ba1fc57e7f7544cea0d1e4bfbce4435c31ab1459)
16. created install registration script and migration
17. [fix: removed lodash and Promise from runtime](https://github.com/oat-sa/extension-tao-itemqti-pci/pull/294/commits/c61a36f945af7d07d5554c887929e7ef01ab5684)

## How to test:

- Run tests in `tao/views/build` with `npx grunt connect:test qtiitempcitest`. It should not fail.
- Checkout this branch and run taoUpdate. (IMS PCI) - Audio recording - v1.0.0 version should be registered and should work correctly.
- Install this branch directly. PCI 1.0.0 version should be registered and should work correctly.
- Check creator part and modify interaction parameters, then run preview mode.
- Create a test with 2 audioRecordingInteraction and check the factory works correctly and they do not listen to other's events.
- Check if you move to another item and move back to this interaction, the response is saved correctly and restored correctly.
- Check the same in new test runner(install extension https://github.com/oat-sa/extension-tao-test-preview-ui-loader). In new test runner, open overview that destroys interaction and check the state restore when you close it. 

image
image

### CODE REVIEW

**Review by marpesia — APPROVED** [review]
Checked on AUT, Preview, Test Runner and Result page.
Good job!

**Review Checklist**
- [x] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful
- [ ] Pull request's target is not `master`
- [ ] Commits are following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ] Commits messages are meaningful
- [ ] Commits are atomic
- [ ] Changelog is updated according to changes (if applicable)
- [ ] Documentation is updated according to changes (if applicable)

**Review by krampstudio — COMMENTED** [review]
Everything seems to work perfectly on the auhtoring side! 
The test runs successfully as well. 

I've a few comments and suggestions on the code: 
 - on the `runtime` side, I think we can easily get rid of 2 dependencies, with Solar the PCI will be sent during the publishing, so if we can easily get rid of optional dependency it would be great:
    -  `lodash`: it looks like it's used only for very basic operations that can easily be replaced by their native alternatives (you can keep it on the creator side)
   - the promise polyfill, considering we don't support IE anymore and the other features have not polyfills, we can just remove it.  
 - Optionally, maybe it's also a good opportunity to update a bit the source code of the creator part to match our coding standards (only the safe changes, as usual)

**Review by Silvia-loza — APPROVED** [review]
All works as expected and unit tests pass fine. Great Job! 
- [X] New code is covered by tests (if applicable)
- [X] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [X] New code is respecting code style rules
- [X] New code is respecting best practices
- [X]  Pull request's target is not` master`
- [X] Feature is working correctly on my local machine (if applicable)
- [X] Acceptance criteria are respected
- [X] Pull request title and description are meaningful

File: views/js/pciCreator/ims/audioRecordingInteraction/creator/widget/states/states.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2017-2022 (original work) Open Assessment Technologies SA;
+ */
+define([
+    'taoQtiItem/qtiCreator/widgets/states/factory',
+    'taoQtiItem/qtiCreator/widgets/interactions/customInteraction/states/states',
+    'audioRecordingInteraction/creator/widget/states/Question'
```
[the remark this answers is not in the document]
Comment (bziondik) [reply views/js/pciCreator/ims/audioRecordingInteraction/creator/widget/states/states.js:21]: This dependency used as `arguments` in line
`return factory.createBundle(states, arguments, ['answer', 'correct', 'map']);`

**Comment by bziondik** [comment]
> Everything seems to work perfectly on the auhtoring side! The test runs successfully as well.
> 
> I've a few comments and suggestions on the code:
> 
> * on the `runtime` side, I think we can easily get rid of 2 dependencies, with Solar the PCI will be sent during the publishing, so if we can easily get rid of optional dependency it would be great:
>   
>   * `lodash`: it looks like it's used only for very basic operations that can easily be replaced by their native alternatives (you can keep it on the creator side)
>   * the promise polyfill, considering we don't support IE anymore and the other features have not polyfills, we can just remove it.
> * Optionally, maybe it's also a good opportunity to update a bit the source code of the creator part to match our coding standards (only the safe changes, as usual)

@krampstudio  fixed In `runtime` - removed `lodash` and promise `polyfill`.
About `creator` part - what is allowed to change? I know that `portableelement`  grunt command didn't support `babel: true,` configs as `bundle` command.
So can I safely use ES2015+ style ?

**Comment by krampstudio** [comment]
> @krampstudio fixed In `runtime` - removed `lodash` and promise `polyfill`. About `creator` part - what is allowed to change? I know that `portableelement` grunt command didn't support `babel: true,` configs as `bundle` command. So can I safely use ES2015+ style ?

You're correct, I have forgotten the creator part was also compiled

**Comment by krampstudio** [comment]
I didn't have retested it but the changes looks good. If it works please merge it.

**Pull request merged**